zephyr/drivers/can/can_sccan.c

```diff
@@ -0,0 +1,1032 @@
+	struct sc_can_data *data = dev->data;
+	struct can_timing timing;
+	int32_t ret;
+	uint32_t v;
+
+	k_mutex_init(&data->inst_mutex);
+	k_sem_init(&data->tx_sem, SCCAN_TX_FIFO_DEPTH, SCCAN_TX_FIFO_DEPTH);
+
+	data->tx_head = 0;
+	data->tx_tail = 0;
+
+	memset(data->rx_cb, 0, sizeof(data->rx_cb));
+
+	data->state = CAN_STATE_STOPPED;
+	data->state_change_cb = NULL;
+
+	/* Set timing according to dts default setting */
+	timing.sjw = config->sjw;
+	ret = can_calc_timing(dev, &timing, config->bus_speed, config->sample_point);
+	if (ret == -EINVAL) {
+		LOG_ERR("Can't find timing for given param");
+		return -EIO;
+	}
+	sc_can_set_timing(dev, &timing);
+
+	/* Enable all IRQ */
+	sc_can_write32(config, SCCAN_IER_OFFSET,
+			  SCCAN_BUSOFFENB |
+			  SCCAN_ACKERENB |
+			  SCCAN_BITERENB |
+			  SCCAN_STFERENB |
+			  SCCAN_FMERENB |
+			  SCCAN_CRCERENB |
+			  SCCAN_RXFOVFENB |
+			  SCCAN_RXFUDFENB |
+			  SCCAN_RXFVALENB |
+			  SCCAN_RCVDNENB |
+			  SCCAN_TXFOVFENB |
+			  SCCAN_TXHBOVFENB |
+			  SCCAN_ARBLSTENB |
+			  SCCAN_TRNSDNENB);
+	config->irq_init(dev);
+
+	/* Dump Version information */
+	v = sc_can_read32(config, SCCAN_VER_OFFSET);
+	LOG_DBG("Space Cubics CAN controller v%d.%d.%d initialized",
+		   SCCAN_VER_MAJOR(v), SCCAN_VER_MINOR(v), SCCAN_VER_PATCH(v));
+
+	return 0;
+}
+
+static int sc_can_get_core_clock(const struct device *dev, uint32_t *rate)
+{
+	const struct sc_can_cfg *config = dev->config;
+
+	*rate = config->clock_frequency;
+
+	return 0;
+}
+
+/*
+ * This API specification is here:
+ *   Get the maximum standard (11-bit) CAN ID filters if false, or extended (29-bit)
+ *   CAN ID filters if true
+ * Acceptance filter of SC CAN Controller supports both standard and extended, so
+ * always returns the same value.
+ */
+static int sc_can_get_max_filters(const struct device *dev, bool ide)
+{
+	ARG_UNUSED(ide);
+
+	return CONFIG_CAN_SCCAN_MAX_FILTER;
+}
+
+static int sc_can_get_max_bitrate(const struct device *dev, uint32_t *max_bitrate)
+{
+	const struct sc_can_cfg *config = dev->config;
+
+	*max_bitrate = config->max_bitrate;
+
+	return 0;
+}
+
+static const struct can_driver_api sc_can_driver_api = {
+	.get_capabilities = sc_can_get_capabilities,
+	.start = sc_can_start,
+	.stop = sc_can_stop,
+	.set_mode = sc_can_set_mode,
+	.set_timing = sc_can_set_timing,
+	.send = sc_can_send,
+	.add_rx_filter = sc_can_add_rx_filter,
+	.remove_rx_filter = sc_can_remove_rx_filter,
+	.get_state = sc_can_get_state,
+#ifndef CONFIG_CAN_AUTO_BUS_OFF_RECOVERY
+	.recover = sc_can_recover,
+#endif
+	.set_state_change_callback = sc_can_set_state_change_callback,
+	.get_core_clock = sc_can_get_core_clock,
+	.get_max_filters = sc_can_get_max_filters,
+	.get_max_bitrate = sc_can_get_max_bitrate,
+	.timing_min = {
+		.sjw = 0x1,
+		.prop_seg = 0x00,
+		.phase_seg1 = 0x04,
+		.phase_seg2 = 0x02,
+		.prescaler = 0x01
+	},
+	.timing_max = {
+		.sjw = 0x4,
+		.prop_seg = 0x00,
+		.phase_seg1 = 0x10,
+		.phase_seg2 = 0x08,
+		.prescaler = 0x400
```

Comment:
How can I verify these values are OK?

Reply:
I'm sorry  this parameter copied from other drivers.
I added the comment and modify the parameters.